### PR TITLE
fix(oca-gen-addon-readme): reduce unneeded merge conflicts

### PR DESCRIPTION
**Pull request by yajo**
As seen in https://github.com/OCA/oca-addons-repo-template/pull/190#issuecomment-1784076360, when somebody tried to merge 2 PRs for the same module, chances are that both imply different readme digests and the merge fails.

As seen in https://github.com/OCA/oca-addons-repo-template/pull/220#issuecomment-1786814130, this was already happening under some circumstances with `requirements.txt` files.

Here I added a `.gitattributes` file which specifies a `merge=union` driver for those files. [Git docs][1] explain that this driver will just keep the changes from all files, marking the merge as succeeded and not including merge markers. This is not ideal, but is OK for our current situation because:

1. Duplicate/conflicting lines in those files should not really harm a lot.
2. The bot will re-render the README upon merge anyways, fixing that conflict automatically.
3. We could define a custom merge driver, but `union` is built into Git, meaning less configurations required downstream.

Still, it doesn't seem to work in octopus merges. At least, if you do consecutive merges instead, all will work as expected.

At the same time, I moved the hook to be the last. This is because previous hooks can reformat the files used to obtain the digest.

Closes #220.

[1]: https://git-scm.com/docs/gitattributes.html#Documentation/gitattributes.txt-union

### CODE REVIEW

File: src/.pre-commit-config.yaml.jinja
Comment (legalsylvain) [inline src/.pre-commit-config.yaml.jinja:232]: not sure to understand this syntax.

**Comment by legalsylvain** [comment]
Thanks a lot for your investigation and your PR ! I'll try this code in the oca days.

File: .github/workflows/test.yml
Comment (pedrobaeza) [inline .github/workflows/test.yml:53]: Why this change?

File: src/.pre-commit-config.yaml.jinja
Comment (yajo) [reply src/.pre-commit-config.yaml.jinja:232]: It's reusing the anchor defined above. Check this out: https://www.linode.com/docs/guides/yaml-anchors-aliases-overrides-extensions/

File: .github/workflows/test.yml
Comment (yajo) [reply .github/workflows/test.yml:53]: The CI failures for previous commit are weird. I don't get those testing locally. I was wondering if maybe pre-commit or pyyaml were too old. Or maybe if a more modern poetry isn't installing dev dependencies, as older poetry did. They had quite a big change in [dependency groups management](https://python-poetry.org/docs/managing-dependencies#dependency-groups).

**Comment by yajo** [comment]
Hi there. This one is ready to merge.

I was having problems because the tests were run under python 3.8 but pre-commit needed python 3.6. I had to pin an older virtualenv version because the latest one doesn't support python 3.6 anymore.

Long story short, now we don't issue a matrix of builds. Instead, we install all required python interpreters in a single build and parallelize at pytest level with pytest-xdist. Thus, now there's just one "test" job.

Since there's just that job, the CI checks will stay as "Required statuses must pass before merging" because I changed the way CI is handled here.

If nobody's against this, I'll merge it so we can unblock the situation explained at https://github.com/OCA/oca-addons-repo-template/pull/220#issuecomment-1802291979.

**Review by pedrobaeza — APPROVED** [review]

**Comment by sbidoul** [comment]
@legalsylvain have you tested this?

**Comment by sbidoul** [comment]
I'm proposing using https://github.com/OCA/maintainer-tools/pull/588 as an alternative.

**Comment by yajo** [comment]
I still think this is a better solution.

Multiple PRs can modify instructions in the readme folder. If those conflict, there's no other solution than to solve it. But if the conflict happens in the README, we shouldn't care when merging, as the bot would re-trigger the generation after merge.

Also this PR refactors the test suite in a more future-proof way (that can be extracted separately, of course).

So IMHO this still has its benefits and we should merge this.

**Comment by sbidoul** [comment]
> Also this PR refactors the test suite in a more future-proof way (that can be extracted separately, of course).

:+1: for that

Regarding the union merge, I think it may be fine for requirements.txt (although I have not personally witnessed such conflicts in practice).

For README.rst/index.hml, I'm not sure.  The drawbacks you mention in the OP don't seem to be worth the gain.

**Comment by yajo** [comment]
OK, nevermind. We can come back to this later if needed.

**Pull request closed without merging**
